File: PanelDX/LeAction_SysF/rbac/contract_access.py
"""Middleware de acesso por status de contrato — consulta em tempo real ao PostgreSQL."""

from __future__ import annotations

from flask import jsonify, request
from psycopg2.extras import RealDictCursor

from rbac.constants import ROLE_SYSADMIN
from rbac.context import resolve_rbac_context
from services.crm_engine import is_contract_access_allowed

_CONTRACT_EXEMPT_PREFIXES = (
    "/status",
    "/gatekeeper",
    "/manutencao",
    "/api/public/",
    "/api/webhooks/",
    "/api/login",
    "/api/auth/",
    "/api/cadastro",
    "/api/verificar-email",
    "/api/admin/crm",
    "/api/admin/cms",
    "/api/admin/esim",
    "/api/admin/usuarios",
    "/api/admin/mesas-inovacao",
    "/integrations/esim",
    "/inovador",
)


def _is_contract_exempt(path: str) -> bool:
    if request.method == "OPTIONS":
        return True
    for prefix in _CONTRACT_EXEMPT_PREFIXES:
        if path == prefix or path.startswith(prefix):
            return True
    return False


def crm_status_contrato_cliente(cursor, id_clie: int) -> str | None:
    """Contrato vigente do cliente — prioriza status operacional mais recente."""
    cursor.execute(
        """
        SELECT status
        FROM public.dx_contratos
        WHERE id_clie = %s
        ORDER BY
            CASE status
                WHEN 'ativo' THEN 0
                WHEN 'trial' THEN 1
                WHEN 'inadimplente' THEN 2
                WHEN 'cancelado' THEN 3
                ELSE 4
            END,
            data_inicio DESC,
            id DESC
        LIMIT 1;
        """,
        (id_clie,),
    )
    row = cursor.fetchone()
    if not row:
        return None
    return row.get("status") if isinstance(row, dict) else row[0]


def register_contract_access_middleware(flask_app) -> None:
    """Bloqueia API quando contrato do id_clie está inadimplente ou cancelado."""

    @flask_app.before_request
    def _enforce_contract_access():
        if _is_contract_exempt(request.path):
            return None

        ctx = resolve_rbac_context()
        if ctx.system_role == ROLE_SYSADMIN:
            return None

        id_clie = ctx.id_clie
        if not id_clie:
            return None

        from app import get_db_conn

        conn = get_db_conn()
        cur = conn.cursor(cursor_factory=RealDictCursor)
        try:
            status = crm_status_contrato_cliente(cur, int(id_clie))
        except Exception:
            cur.close()
            return None
        finally:
            cur.close()

        if is_contract_access_allowed(status):
            return None

        return jsonify({
            "success": False,
            "status": "error",
            "error": "Acesso suspenso: contrato inadimplente ou cancelado.",
            "contract_status": status,
            "code": "CONTRACT_BLOCKED",
        }), 403

File: mudaedu/LeAction_Sys_FE/middleware/contract-access.js
'use strict';

const { isSessionAdmin } = require('../lib/auth-session');
const {
    fetchContractStatusForClie,
    isContractAccessAllowed,
    resolveIdClieForContract,
} = require('../lib/contract-access');
const { buildActionHubCheckoutUrl } = require('../lib/actionhub-checkout');

const CONTRACT_EXEMPT_PATHS = new Set([
    '/',
    '/cadastro',
    '/login',
    '/logout',
    '/termos-de-uso',
    '/instrucoes-de-uso',
    '/versao-aplicacao',
    '/verificar-email',
    '/consultor-leaction',
    '/portal-consultor',
    '/manutencao',
    '/gatekeeper/bypass',
    '/gatekeeper/unlock',
    '/gatekeeper/lock',
]);

const CONTRACT_EXEMPT_PREFIXES = [
    '/css/',
    '/js/',
    '/images/',
    '/img/',
    '/gatekeeper/',
    '/gatekeeper',
    '/api/public/',
    '/api/admin/crm',
    '/bff/admin/crm',
    '/bff/admin/consultores',
    '/bff/consultor',
    '/admin/crm',
    '/admin',
    '/admin-cms',
    '/admin/cms',
    '/admin/esim',
    '/admin/usuarios',
    '/admin/mesas-inovacao',
];

function isContractExempt(pathname) {
    if (CONTRACT_EXEMPT_PATHS.has(pathname)) return true;
    return CONTRACT_EXEMPT_PREFIXES.some(
        (prefix) => pathname === prefix || pathname.startsWith(prefix)
    );
}

function isElbHealthCheck(req) {
    const ua = String(req.headers['user-agent'] || '');
    return ua.includes('ELB-HealthChecker');
}

/**
 * Bloqueia navegação quando contrato do cliente está inadimplente ou cancelado.
 * Consulta o banco a cada requisição (reflete alterações imediatamente).
 */
async function contractAccessMiddleware(req, res, next) {
    if (isContractExempt(req.path) || isElbHealthCheck(req)) {
        return next();
    }

    if (!req.session || (!req.session.lead && !req.session.isTeam && !req.session.user_id)) {
        return next();
    }

    if (isSessionAdmin(req)) {
        return next();
    }

    let idClie;
    try {
        idClie = await resolveIdClieForContract(req);
    } catch (err) {
        console.warn('[ContractAccess] Falha ao resolver id_clie:', err.message);
        return next();
    }

    if (!idClie) {
        return next();
    }

    let status;
    try {
        status = await fetchContractStatusForClie(idClie);
    } catch (err) {
        const missingTable = /dx_contratos|relation.*does not exist/i.test(err.message || '');
        if (missingTable) {
            return next();
        }
        console.error('[ContractAccess] Falha ao consultar contrato:', err.message);
        return next();
    }

    if (isContractAccessAllowed(status)) {
        return next();
    }

    const wantsJson =
        req.xhr ||
        (req.headers.accept && req.headers.accept.indexOf('json') > -1) ||
        req.path.startsWith('/api/') ||
        req.path.startsWith('/bff/');

    const payload = {
        success: false,
        error: 'Acesso suspenso: contrato inadimplente ou cancelado.',
        contract_status: status,
        code: 'CONTRACT_BLOCKED',
    };

    if (wantsJson) {
        return res.status(403).json(payload);
    }

    return res.status(403).render('error', {
        title: 'Acesso suspenso',
        message: 'O contrato da sua empresa está inadimplente ou cancelado. Renove ou faça upgrade para continuar usando o MudaEdu.',
        contractStatus: status,
        checkoutUpgradeUrl: buildActionHubCheckoutUrl(req, idClie, {
            returnTo: req.path || '/projeto',
        }),
        statusCode: 403,
    });
}

module.exports = {
    contractAccessMiddleware,
    isContractExempt,
};
